src/openrct2/core/OrcaStream.hpp

```diff
@@ -0,0 +1,679 @@
+                    return static_cast<T>(raw);
+                }
+                else
+                {
+                    if constexpr (std::is_signed<T>())
+                    {
+                        int32_t raw{};
+                        Read(&raw, sizeof(raw));
+                        if (raw < std::numeric_limits<T>::min() || raw > std::numeric_limits<T>::max())
+                        {
+                            throw std::runtime_error("Value is incompatible with internal type.");
+                        }
+                        return static_cast<T>(raw);
+                    }
+                    else
+                    {
+                        uint32_t raw{};
+                        Read(&raw, sizeof(raw));
+                        if (raw > std::numeric_limits<T>::max())
+                        {
+                            throw std::runtime_error("Value is incompatible with internal type.");
+                        }
+                        return static_cast<T>(raw);
+                    }
+                }
+            }
+
+            template<typename T, typename = std::enable_if<std::is_integral<T>::value>> void WriteInteger(const T value)
+            {
+                if constexpr (sizeof(T) > 4)
+                {
+                    if constexpr (std::is_signed<T>())
+                    {
+                        auto raw = static_cast<int64_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                    else
+                    {
+                        auto raw = static_cast<uint64_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                }
+                else
+                {
+                    if constexpr (std::is_signed<T>())
+                    {
+                        auto raw = static_cast<int32_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                    else
+                    {
+                        auto raw = static_cast<uint32_t>(value);
+                        Write(&raw, sizeof(raw));
+                    }
+                }
+            }
+
+            std::string ReadString()
+            {
+                std::string buffer;
+                buffer.reserve(64);
+                while (true)
+                {
+                    char c;
+                    ReadBuffer(&c, sizeof(c));
+                    if (c == 0)
+                    {
+                        break;
+                    }
+                    buffer.push_back(c);
+                }
+                buffer.shrink_to_fit();
+                return buffer;
+            }
+
+            void WriteString(const std::string_view s)
+            {
+                const char nullt = '\0';
+                auto len = s.find('\0');
+                if (len == std::string_view::npos)
+                {
+                    len = s.size();
+                }
```

Comment:
```suggestion
                const char nullt = '\0';
                auto len = s.size();
                while (len > 0 && s[len - 1] == nullt)
                {
                    len--;
                }
```
I think in 99.9% of the time this will simply skip 1 byte, not sure if this is cleaner to read but we don't have to search for null terminators at the beginning.

Reply:
If you ever pass a `std::string_value` containing null terminators (which could even be the first byte)... it will break reading the string when opening the file. It has to ensure no more than one null terminator byte is written.

Reply:
I don't think we have such strings, if we do it would be perhaps best to encode the size as well and not treat them as strings

Reply:
I’m a bit reluctant to just go on an _assumption_ we don’t have such strings. Much prefer having more robust code.

Reply:
I also prefer robust code, thats why I think meddling with null termination on string_views is a terrible idea, string_views give you zero guarantees about null termination. The point of string_view is to use the size provided and not rely on termination.

Reply:
The `string_view` is just an overload. You also have `char*` and `std::string` overloads. Regardless of the overload, they all need to comply with the file format.

Reply:
Well my point is somewhat that you should not pass data with string_view/string/char* or by any other means that is not an actual string or includes multiple null terminators, this only means the API is used incorrectly. Anyway its just a nitpick and would reduce the overall lookup time when it comes to string encoding.